lib/mss/mss_parser.js

```diff
@@ -0,0 +1,1050 @@
+    }
+
+    const variants = [];
+    for (const audio of (audioStreams.length > 0 ? audioStreams : [null])) {
+      for (const video of (videoStreams.length > 0 ? videoStreams : [null])) {
+        variants.push(this.createVariant_(audio, video));
+      }
+    }
+    context.variants = variants;
+    context.textStreams = textStreams;
+  }
+
+  /**
+   * @param {!Element} streamIndex
+   * @param {!Element} qualityLevel
+   * @param {!Array.<shaka.mss.MssParser.TimeRange>} timeline
+   * @param {!Array.<shaka.extern.DrmInfo>} drmInfos
+   * @param {!shaka.mss.MssParser.Context} context
+   * @return {?shaka.extern.Stream}
+   * @private
+   */
+  createStream_(streamIndex, qualityLevel, timeline, drmInfos, context) {
+    const XmlUtils = shaka.util.XmlUtils;
+    const ContentType = shaka.util.ManifestParserUtils.ContentType;
+    const MssParser = shaka.mss.MssParser;
+
+    const type = streamIndex.getAttribute('Type');
+    const isValidType = type === 'audio' || type === 'video' ||
+        type === 'text';
+    if (!isValidType) {
+      shaka.log.alwaysWarn('Ignoring unrecognized type:', type);
+      return null;
+    }
+
+    const lang = streamIndex.getAttribute('Language');
+    const id = this.globalId_++;
+
+    const bandwidth = XmlUtils.parseAttr(
+        qualityLevel, 'Bitrate', XmlUtils.parsePositiveInt);
+    const width = XmlUtils.parseAttr(
+        qualityLevel, 'MaxWidth', XmlUtils.parsePositiveInt);
+    const height = XmlUtils.parseAttr(
+        qualityLevel, 'MaxHeight', XmlUtils.parsePositiveInt);
+    const channelsCount = XmlUtils.parseAttr(
+        qualityLevel, 'Channels', XmlUtils.parsePositiveInt);
+    const audioSamplingRate = XmlUtils.parseAttr(
+        qualityLevel, 'SamplingRate', XmlUtils.parsePositiveInt);
+
+    /** @type {!shaka.extern.Stream} */
+    const stream = {
+      id: id,
+      originalId: streamIndex.getAttribute('Name') || String(id),
+      createSegmentIndex: () => Promise.resolve(),
+      closeSegmentIndex: () => Promise.resolve(),
+      segmentIndex: null,
+      mimeType: '',
+      codecs: '',
+      frameRate: undefined,
+      pixelAspectRatio: undefined,
+      bandwidth: bandwidth || 0,
+      width: width || undefined,
+      height: height || undefined,
+      kind: '',
+      encrypted: drmInfos.length > 0,
+      drmInfos: drmInfos,
+      keyIds: new Set(),
+      language: lang || 'und',
+      label: '',
+      type: '',
+      primary: false,
+      trickModeVideo: null,
+      emsgSchemeIdUris: [],
+      roles: [],
+      forced: false,
+      channelsCount: channelsCount,
+      audioSamplingRate: audioSamplingRate,
+      spatialAudio: false,
+      closedCaptions: null,
+      hdr: undefined,
+      tilesLayout: undefined,
+      matchedStreams: [],
+      mssPrivateData: {
+        duration: context.duration,
+        timescale: context.timescale,
+        codecPrivateData: null,
+      },
+    };
+
+    const subType = streamIndex.getAttribute('Subtype');
```

Comment:
Please add a comment that this is specifically for text tracks.

Reply:
Addded!